core/src/main/java/org/apache/calcite/plan/RelOptPredicateList.java

```diff
@@ -233,7 +234,8 @@ public boolean isEffectivelyNotNull(RexNode e) {
         return true;
       }
     }
-    if (SqlKind.COMPARISON.contains(e.getKind())) {
+    // Skip IN RexSubQuery because it is not literal we want to handle here
+    if (SqlKind.COMPARISON.contains(e.getKind()) && !(e instanceof RexSubQuery)) {
```

Comment:
Would it be perhaps more clear and flexible to express the new condition in a more generic way? (just in case at some point in time this exceptional "1 operand comparison" is not limited to RexSubQuery), e.g.:
```
    if (SqlKind.COMPARISON.contains(e.getKind())) {
      // A comparison with a (non-null) literal, such as 'ref < 10', is not null if 'ref'
      // is not null.
      List<RexNode> operands = ((RexCall) e).getOperands();
      // We can have one operand in case e.g. of a RexSubQuery with IN operator.
      if (operands.size() > 1 && operands.get(1) instanceof RexLiteral
          && !((RexLiteral) operands.get(1)).isNull()) {
        return isEffectivelyNotNull(operands.get(0));
      }
    }
```

Reply:
This is more clear and can handle more illegal cases besides RexSubQuery, I will change it.

Reply:
@rubenada updated.